bot/utils/persistent_scheduling.py

```diff
@@ -0,0 +1,155 @@
+SchedulerTaskFactory = t.Callable[[CacheKey], t.Coroutine]
+
+
+class PersistentScheduler:
+    """
+    Schedule the execution of coroutines and keep track of them even across restarts and cog reloads.
+
+    When instantiating a PersistentScheduler, a name must be provided. This name is used to distinguish the
+    instance's log messages and RedisCache from other instances. Using the name of the class or module containing
+    the instance is suggested. The same name must be provided after a restart for the tasks to be preserved.
+
+    A coroutine must be provided in order to reschedule each of the keys contained in the cache at init.
+    The coroutine will take a single argument -- the ID that was used to create the Scheduler task in the first place.
+
+    Coroutines can be scheduled with `schedule_at` or `schedule_later`.
+    A unique ID is required to be given in order to keep track of the resulting Tasks.
+    Any scheduled task can be cancelled prematurely using `cancel` by providing the same ID used to schedule it, but
+    the task will be rescheduled from the cache on the next initialization. To remove the task completely, you can
+    `delete` it.
+    The `in` operator is supported for checking if a task with a given ID is currently scheduled.
+
+    Any exception raised in a scheduled task is logged when the task is done.
+    """
+
+    def __init__(self, name: str, task_factory: SchedulerTaskFactory, event_loop: AbstractEventLoop):
+        self._scheduler = Scheduler(name)
+        self.cache = RedisCache(namespace=name)
+
+        self.task_factory = task_factory
+
+        self._reschedule_task = event_loop.create_task(self.reschedule_all_tasks())
+
+    def __contains__(self, task_id: CacheKey) -> bool:
+        """Return True if a task with the given `task_id` is currently scheduled."""
+        return task_id in self._scheduler
+
+    async def wait_until_ready(self) -> None:
+        """Wait until the scheduler is ready and the cache is ready (and all cached tasks are scheduled)."""
+        await self._reschedule_task
+
+    async def get(self, task_id: CacheKey) -> t.Optional[Arrow]:
+        """Get the time in which `task_id`'s task should go off. If the task is not found return None."""
+        timestamp = await self.cache.get(task_id)
+
+        if not timestamp:
+            return None
+        return Arrow.utcfromtimestamp(timestamp)
+
+    async def to_dict(self) -> t.Dict[CacheKey, Arrow]:
+        """
+        Get a dict representation of the currently scheduled tasks.
+
+        The dictionary contains the task_id's as keys, and the times in which the tasks should go off as values.
+        """
+        dict_ = await self.cache.to_dict()
+        return {task_id: Arrow.utcfromtimestamp(timestamp) for task_id, timestamp in dict_.items()}
+
+    async def reschedule_task(self, task_id: CacheKey) -> None:
+        """Reschedule the task according to `task_factory` and `task_id`."""
```

Comment:
Please document what happens if the task is already scheduled.

Reply:
On second thought, what is even the use case of this? The task wouldn't exist in the cache unless it's already scheduled. Furthermore, `reschedule_all_tasks` takes care of rescheduling it on restart.